### ISSUE TITLE
autoreloading weapons is not picky about ammo fullness

### ISSUE THREAD

**Issue by lynxlynxlynx** [ja2-stracciatella/ja2-stracciatella#685]
When you run out of ammo, you can click as if shooting for the merc to reload it. However, if you have several magazines, the merc doesn't necessarily do the smart thing.

Eg. if you have two 20 pack 7.62mm magazines, one full and one with 5 bullets remaining, the soldier may pick the second one. Quite a bad choice in the middle of battle.

**Comment by DSeyka** [comment]
It's better for inventory management!
Seriously though I'm pretty sure vanilla game just takes the first thing it finds in the inventory which can be used to reload the weapon currently in hands. And as far as stacked clips/magazines are concerned, first thing is the one "on top" of others, be it full or nearly empty.

**Comment by lynxlynxlynx** [comment]
Good point, also the mag type matters. I guess we should look up all appropriate ammo, then select the mags with the same capacity as the weapon (if any), then select the fullest.

Maybe we could also distinguish by bullet type (-, ap, hp), to try to reload with the same ammo, but I don't think it really matters.

**Comment by DSeyka** [comment]
If this is implemented it should be more abstract, i.e. taking any modded ammo/weapons into account.

**Comment by lynxlynxlynx** [comment]
That's why I said I don't think it matters. First it's a separate item and second, who decides on the priority between type and fullness?

**Comment by DSeyka** [comment]
Well, magazines and clips are virtually unlimited: you get them by unloading a weapon of corresponding type no matter how many bullets you have left in it. You can effectively transmute pistol clips and SMG magazines in the game even though they have different weights and stackability.

So it's probably better to base the whole thing on fullness, e.g. reload with whatever is available so the gun as full much as we can with a single reload action? I pretty much think it should be consistent more than anything else.

**Comment by lynxlynxlynx** [comment]
The only case that would fit within the initial issue is doing an extra check in case you have several types of big mags to choose from and they're all full.

**Comment by JAsmine-ja2** [comment]
One thing to consider: When reloading the gun in TB mode, reloading from an unfitting magazine costs more AP (e.g. loading a .45 Colt pistol with the .45 SMG magazine).

**Comment by RinneT** [comment]
A couple of things:
I am 99% certain the game already takes into account the ammo type (-, hp, ap).
I am 95% certain the game already takes into account magazine size.

Also taking the first magazine from the stack always felt like a design choice to me.
In real life, how long would it take you to search through your stuff to find the fullest magazine? In the game you either take that task off your mercs, or they'll pick the top one.
To me a well organized inventory is part of the battle preparation.

Only thing I'm unsure of is which inventory slot order they'll use.

All in all: Probably a good change for many players, but please make sure to use a json switch :)

**Comment by lynxlynxlynx** [comment]
In real life, you wouldn't be carrying around near empty mags.
And to me, a well organised inventory is one where I can see if any mags aren't full, so they have to be on the top of the stack (in the visibility sense; it's just a gui problem). That's part of the battle prep.

**Comment by JAsmine-ja2** [comment]
> a well organised inventory is one where I can see if any mags aren't full, so they have to be on the top of the stack (in the visibility sense; it's just a gui problem). That's part of the battle prep.
+1 on this

To be honest, I never use the "auto-reload" feature because of this. I really don't consider this to be a game-changing / advantage thing for the player. Because of my experience with the evolution of v1.13, I'm 100% pro using a switch for this.

**Comment by DSeyka** [comment]
>I am 95% certain the game already takes into account magazine size (i.e. smg vs. pistol).

This needs testing, I was under impression that it doesn't care about magazine size and just takes whatever it finds first in the inventory.

**Comment by Manuel-K** [comment]
> a well organised inventory is one where I can see if any mags aren't full, so they have to be on the top of the stack

**+2**

The item stacking makes inventory management very annoying. IMO it's one of the game's worst UI problems.

**Comment by lynxlynxlynx** [comment]
@DSeyka it's true, at least in turn-based as previously noted (requiring more ap).

**Comment by DSeyka** [comment]
So, putting a pistol clip into an SMG requires less AP than an SMG one?
I'd totally go for less ammo for less APs in a heated firefight, it's a tactical choice then. Not that it's even remotely realistic. :)
On the other hand it's the same in vanilla.

**Comment by lynxlynxlynx** [comment]
No, the other way around, since the clip doesn't fit and you have to feed it bullet by bullet. It just checks whether the ammo format is what the weapon uses.

**Comment by DSeyka** [comment]
Well, that makes much more sense now. But I guess the AP cost is negligible anyway? Because I honestly never noticed that, but maybe that's because a reload is an end of turn 99% of time anyway.

**Comment by chmallon** [comment]
`AutoReload()` uses `FindAmmoToReload()` to get a magazine.
This in turn does up to three passes to find ammunition in the following order:
1. Find same ammunition type (calibre, mag size and bullet type);
1. Find ammunition with same calibre and clip size, ignoring bullet type;
1. Find ammunition with same calibre, ignoring clip size and ignoring bullet type.

The first successful pass stops the search.
In each pass the inventory slots are iterated from `HELMET` to `SMALLPOCK8POS`.
The first slot with ammunition meeting the criteria of the pass stops the pass.

If there are multiple clips in the found slot, then the first one in the stack is used.
So you can steer which clip will be used:
Put the desired clip in the smallest slot (according to the inventory order) and there in the first place of a stack.

**Comment by JAsmine-ja2** [comment]
The AutoReload system is okay the way it is.
A simple (?) enhancement would be to leave the whole thing intact and just allow the player to select (json switch) which mag in the stack is used.
1st one (vanilla) or last one (new)

**Comment by lynxlynxlynx** [comment]
That potentially does nothing. The original problem description still stands — no change or pick the full one, if any.

The discussion about mags/clips was resolved, while the one about bullet type would just complicate matters too much for such a rare chance to be useful.

**Comment by JAsmine-ja2** [comment]
@lynxlynxlynx So, if I get you right, you propose that the current routine
> 1. Find same ammunition type (calibre, mag size and bullet type);
> 2. Find ammunition with same calibre and clip size, ignoring bullet type;
> 3. Find ammunition with same calibre, ignoring clip size and ignoring bullet type.

should be modified towards something like this:
1. Find same ammunition type (calibre, mag size and bullet type) **with full capacity**;
2. Find same ammunition type (calibre, mag size and bullet type) **regardless of remaining capacity**;
3. Find ammunition with same calibre and clip size, ignoring bullet type;
4. Find ammunition with same calibre, ignoring clip size and ignoring bullet type.

?

**Comment by lynxlynxlynx** [comment]
Yep.

**Comment by Manuel-K** [comment]
How should stacks be handled?

IMO the only important change would be to look for the best magazine inside the selected check.

**Comment by lynxlynxlynx** [comment]
This is all about stacks, though I suppose there could be players that carry unstacked ammo.

**Comment by Manuel-K** [comment]
Okay. The original code searches a slot (i.e. a slot) and then takes the first element (data structure wise) from the stack in that slot.
